src/main/scala/KotlinTest.scala

```diff
@@ -1,45 +1,71 @@
+import xsbti.compile.{CompileAnalysis, FileHash, MiniOptions, MiniSetup, SingleOutput}
+import xsbti.{AnalysisCallback, VirtualFile, VirtualFileRef}
 
 object KotlinTest {
   private object EmptyLookup extends Lookup {
-    def changedClasspathHash: Option[Vector[FileHash]] = None
+    override def changedClasspathHash: Option[Vector[FileHash]] = None
 
-    def analyses: Vector[CompileAnalysis] = Vector.empty
+    override def analyses: Vector[CompileAnalysis] = Vector.empty
 
-    def lookupOnClasspath(binaryClassName: String): Option[File] = None
+    override def lookupOnClasspath(binaryClassName: String): Option[VirtualFileRef] = None
 
-    def lookupAnalysis(binaryClassName: String): Option[CompileAnalysis] = None
-    def changedBinaries(previousAnalysis: xsbti.compile.CompileAnalysis): Option[Set[java.io.File]] = None
-    def changedSources(previousAnalysis: xsbti.compile.CompileAnalysis): Option[xsbti.compile.Changes[java.io.File]] = None
-    def removedProducts(previousAnalysis: xsbti.compile.CompileAnalysis): Option[Set[java.io.File]] = None
-    def shouldDoIncrementalCompilation(changedClasses: Set[String],analysis: xsbti.compile.CompileAnalysis): Boolean = true
+    override def lookupAnalysis(binaryClassName: String): Option[CompileAnalysis] = None
+    override def changedBinaries(previousAnalysis: xsbti.compile.CompileAnalysis): Option[Set[VirtualFileRef]] = None
+    override def changedSources(previousAnalysis: xsbti.compile.CompileAnalysis): Option[xsbti.compile.Changes[VirtualFileRef]] = None
+    override def removedProducts(previousAnalysis: xsbti.compile.CompileAnalysis): Option[Set[VirtualFileRef]] = None
+    override def shouldDoIncrementalCompilation(changedClasses: Set[String],analysis: xsbti.compile.CompileAnalysis): Boolean = true
 
-    def hashClasspath(x$1: Array[java.io.File]): java.util.Optional[Array[xsbti.compile.FileHash]] = java.util.Optional.empty()
+    override def hashClasspath(classpath: Array[VirtualFile]): java.util.Optional[Array[FileHash]] = java.util.Optional.empty()
   }
 
   val kotlinTests = Def.task {
+    import sbt.internal.inc.PlainVirtualFileConverter.converter
+    val defaultStampReader = Stamps.timeWrapBinaryStamps(converter)
+
     val out = ((Test / target).value ** "scala-*").get.head / "test-classes"
-    val srcs = ((Test / sourceDirectory).value ** "*.kt").get.toList
-    val xs = (out ** "*.class").get.toList
+    val srcs = ((Test / sourceDirectory).value ** "*.kt").get.map(f => converter.toVirtualFile(f.toPath)).toList
+    val xs = (out ** "*.class").get.map(_.toPath).toList
 
-    val loader = ClasspathUtilities.toLoader((Test / fullClasspath).value map {
-      _.data
-    })
+    val loader = ClasspathUtil.toLoader((Test / fullClasspath).value.map(_.data.toPath))
     val log = streams.value.log
     val output = new SingleOutput {
       def getOutputDirectory: File = out
     }
-    val a0 = IncrementalCompile(
+    val currentSetup = MiniSetup.of(
```

Comment:
About this `MiniSetup`:

- it seems to work in what we test
- many of the options I left empty seem to be for `javac`/`scalac`, or to not be needed for what we do here

But I wouldn't say I'm 100% confident about this piece of code...

Note : the incremental compiler private APIs are used here to provide a `CompileAnalysis`, which is needed to call `Tests.discover`. It's ugly but I assume is why it was done like that because it would be a pain to do this ourselves using only public APIs.

Reply:
FYI: https://github.com/davidB/scala-maven-plugin/blob/master/src/main/java/sbt_inc/SbtIncrementalCompilers.java#L199